File: Android/RecyclerViewLayoutManagerStudy/app/src/main/java/com/android/gold/recyclerviewlayoutmanagerstudy/ListAdapter.java
package com.android.gold.recyclerviewlayoutmanagerstudy;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;

/**
 * Created by Gold on 2017. 9. 21..
 */

public class ListAdapter extends RecyclerView.Adapter<ListAdapter.ViewHolder> {

    ArrayList<Integer> data;

    public ListAdapter(ArrayList<Integer> data) {
        this.data = data;
    }

    @Override
    public ListAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item, null);
        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(ListAdapter.ViewHolder holder, int position) {

    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        public ViewHolder(View itemView) {
            super(itemView);
        }
    }
}

File: Android/GithubUserSearch/app/src/main/java/gold/dev/com/githubusersearch/ListAdapter.kt
package gold.dev.com.githubusersearch

import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import gold.dev.com.githubusersearch.databinding.ListItemBinding
import gold.dev.com.githubusersearch.model.Users
import gold.dev.com.githubusersearch.viewmodel.MainViewModel

class ListAdapter(private val mainViewModel: MainViewModel) : RecyclerView.Adapter<ListAdapter.ViewHolder>() {

    var data: ArrayList<Users.Item> = ArrayList()
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ListAdapter.ViewHolder {
        val binding = DataBindingUtil.inflate<ListItemBinding>(
            LayoutInflater.from(parent.context),
            R.layout.list_item,
            parent,
            false
        )

        return ViewHolder.DefaultViewHolder(binding)
    }

    override fun getItemCount(): Int = data.size

    override fun onBindViewHolder(holder: ListAdapter.ViewHolder, position: Int) = when (holder) {
        is ViewHolder.DefaultViewHolder -> {
            val user = data.getOrNull(position) ?: run { return }
            holder.binding.user = user
            holder.binding.viewModel = mainViewModel
            holder.binding.executePendingBindings()
        }
    }

    sealed class ViewHolder(binding: ViewDataBinding) : RecyclerView.ViewHolder(binding.root) {

        class DefaultViewHolder(val binding: ListItemBinding) : ViewHolder(binding)

    }

}